File: src/day13/solution.py
import sys
import os

day_number = sys.path[0].split('\\')[-1]
if len(sys.argv)==1:
    path_to_source = os.path.join("\\".join(sys.path[0].split("\\")[:-2]), f"input\\{day_number}")
else:
    path_to_source = os.path.join("\\".join(sys.path[0].split("\\")[:-2]), f"sample\\{day_number}")
with open(path_to_source, "r") as infile:
    lines = infile.read().split("\n")

depart = int(lines[0])

ids = []
delays  = []
for i, nums in enumerate(lines[1].split(",")):
    if(nums != 'x'):
        ids.append(int(nums))
        delays.append(i)

best = 1e100
for id in ids:
    delay = (-depart%id)
    if(delay < best):
        best = delay
        part1 = delay*id


t = [0 for _ in range(len(ids))]
prod = [id for id in ids]
for i in range(1, len(ids)):
    prod[i] *= prod[i-1]
    
for i in range(len(ids)-1):
    inv_prod = prod[i]**(ids[i+1]-2) % ids[i+1]
    f = ( (-t[i]-delays[i+1]) * inv_prod ) % ids[i+1]
    t[i+1] = t[i] + f*prod[i]
    
    
part2= t[-1]

print(part1)
print(part2)
        

File: src/day13/Solution.java
package day13;

import static utils.ModuloArithmetic.expMod;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import utils.NoSolutionException;

public class Solution {
  private static final boolean USE_SAMPLE = false;
  private final List<Integer> ids;
  private final List<Integer> delays;
  private final int target;

  Solution(List<String> lines){
    int idx = 0;
    ids = new ArrayList<>();
    delays = new ArrayList<>();
    target = Integer.parseInt(lines.get(0));
    for(String s:lines.get(1).split(",")){
      if(!s.matches("x")){
        ids.add(Integer.parseInt(s));
        delays.add(idx);
      }
      idx++;
    }
  }
  public String partOne() throws NoSolutionException {
    int best = Integer.MAX_VALUE;
    int solution = -1;
    for (int id:ids){
      int delay = Math.floorMod(-target, id);
      best =  best<delay? best:delay;
      solution = best<delay?solution:delay*id;
    }
    return String.valueOf(solution);
  }

  public String partTwo() throws NoSolutionException {
    long[] times = new long[ids.size()];
    long[] prods = new long[ids.size()];
    prods[0] = ids.get(0);
    for(int i = 1; i<ids.size(); i++){
      prods[i] = ((long)ids.get(i))*prods[i-1];
    }
    for(int i = 0; i<times.length-1; i++){
      long inv_prod = expMod(Math.floorMod(prods[i], ids.get(i+1)), ids.get(i+1)-2L, ids.get(i+1));
      long f = Math.floorMod( (-times[i]-delays.get(i+1))*inv_prod, ids.get(i+1));
      times[i+1] = times[i]+f*prods[i];
    }
    return String.valueOf(times[times.length-1]);
  }

  public static void main(String[] args) {
    try{
      List<String> lines = Files.readAllLines(Paths.get((USE_SAMPLE?"./sample/":"./input/")+Solution.class.getPackageName()));
      Solution s = new Solution(lines);
      System.out.println(s.partOne());
      System.out.println(s.partTwo());
    } catch (IOException | NoSolutionException e) {
      e.printStackTrace();
    }
  }
}
